bundles/org.openhab.binding.souliss/src/main/java/org/openhab/binding/souliss/internal/protocol/UDPDecoder.java

```diff
@@ -0,0 +1,581 @@
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class decodes incoming Souliss packets, starting from decodevNet
+ *
+ * @author Tonino Fazio - Initial contribution
+ * @author Luca Calcaterra - Refactor for OH3
+ * @author Alessandro Del Pex - Souliss App
+ */
+@NonNullByDefault
+public class UDPDecoder {
+
+    private final Logger logger = LoggerFactory.getLogger(UDPDecoder.class);
+    @Nullable
+    private DiscoverResult discoverResult;
+    @Nullable
+    private SoulissGatewayHandler gwHandler;
+
+    @Nullable
+    private Byte lastByteGatewayIp = null;
+
+    public UDPDecoder(Bridge bridge, @Nullable DiscoverResult pDiscoverResult) {
+        this.gwHandler = (SoulissGatewayHandler) bridge.getHandler();
+        this.discoverResult = pDiscoverResult;
+        var localGwHandler = this.gwHandler;
+        if (localGwHandler != null) {
+            this.lastByteGatewayIp = (byte) Integer.parseInt(localGwHandler.gwConfig.gatewayLanAddress.split("\\.")[3]);
+        }
+    }
+
+    /**
+     * Get packet from VNET Frame
+     *
+     * @param packet
+     *            incoming datagram
+     */
+    public void decodeVNetDatagram(DatagramPacket packet) {
+        int checklen = packet.getLength();
+        ArrayList<Byte> mac = new ArrayList<>();
+        for (var ig = 7; ig < checklen; ig++) {
+            mac.add((byte) (packet.getData()[ig] & 0xFF));
+        }
+
+        // Check if decoded Gw equal to ip of bridge handler or 1 (action messages)
+        Byte gwCheck = (byte) (packet.getData()[5] & 0xFF);
+        if ((gwCheck == 1) || (gwCheck.equals(this.lastByteGatewayIp))) {
+            decodeMacaco((byte) (packet.getData()[5] & 0xFF), mac);
+        }
+    }
+
+    /**
+     * Decodes lower level MaCaCo packet
+     *
+     * @param lastByteGatewayIP
+     *
+     * @param macacoPck
+     */
+    private void decodeMacaco(byte lastByteGatewayIP, ArrayList<Byte> macacoPck) {
+        int functionalCode = macacoPck.get(0);
+        switch (functionalCode) {
+            case SoulissUDPConstants.SOULISS_UDP_FUNCTION_PING_RESP:
+                logger.debug("Received functional code: 0x{}- Ping answer", Integer.toHexString(functionalCode));
+                decodePing(macacoPck);
+                break;
+
+            case SoulissUDPConstants.SOULISS_UDP_FUNCTION_DISCOVER_GW_NODE_BCAST_RESP:
+                logger.debug("Received functional code: 0x{} - Discover a gateway node answer (broadcast)",
+                        Integer.toHexString(functionalCode));
+                try {
+                    decodePingBroadcast(macacoPck);
+                } catch (UnknownHostException e) {
+                    logger.debug("Error: {}", e.getLocalizedMessage());
+                    logger.error("Error:", e);
```

Comment:
See above. Logging to error. Please check all.

Reply:
fixed all classes